In [2]:
#La base de datos "wine" contiene muestras de 3 clases de vinos. Cada una de las observaciones tiene 13 atributos (ver http://archive.ics.uci.edu/ml/datasets/Wine)
#Encuentra la combinación de estrategia de pre-procesamiento + parámetro de regularización C del clasificador logístico
#con la que se obtenga el mejor acierto de clasificación. Se deben contemplar al menos 5 parámetros de C (suficientemente separados entre sí) 
#y 3 estrategias de pre-procesamiento. 

import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from numpy import genfromtxt #librería para leer archivos de texto
#Primero debes ingresar a la URL del párrafo superior y descargar la base de datos. La encontrarás como wine.data.
#Una vez descargada, guárdala en la misma ruta del notebook


data = genfromtxt('wine.data', delimiter=',') #almacenamos la base de datos en la variable data
print(type(data),data.dtype) #nos muestra el tipo de varible que es data y el tipo de datos que contiene
print(data.shape)
#debemos tener en cuenta que para esta base de datos la primera columna es la etiqueta y las siguientes 13 son los 
#atributos.


<class 'numpy.ndarray'> float64
(178, 14)


** Ejercicio 1 **

Desarrolla un código que almacene en la variable X la matriz de observaciones (para este caso sería una matriz de 178x13) y
en Y las etiquetas (vector de 178 posiciones). Adicionalmente debes convertir el vector de etiquetas en un vector de números enteros

In [3]:
#aquí está el espacio para desarrollar el ejericicio 1

observaciones =[]
#etiquetas2 = []
etiquetas = []

#for x in range (0,177):
#    etiquetas2.append(data[x,0])
etiquetas = data[:,0]   
#print (etiquetas2)
#print (etiquetas)
#print (etiquetas.shape)
observaciones = data[:,1:]
#print(observaciones)
#print(etiquetas.dtype)
X = observaciones
Y = etiquetas.astype(int)
print(X.shape)
print(Y.shape)


(178, 13)
(178,)


** Ejercicio 2 **

1. Divide la matriz de observaciones en 80% para entrenamiento y 20% para validación
2. Genera las 3 estrategias de preprocesamiento y aplícalas correctamente a las matrices X_train y X_test
3. Instancia los 5 clasificadores (1 para cada valor de C)
4. Entrena cada clasificador para cada una de las estrategias de preprocesamiento
5. Valida los clasificadores con cada uno de los X_test
6. Concluye. ¿Cuál fue la mejor combinación de C y preprocesamiento?

In [4]:
#espacio para solucionar el ejercicio 2
#punto 1
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size=0.2, random_state=0)

print(X_train.shape)
print(np.shape(X_test)[0], np.shape(X_test)[1])

(142, 13)
36 13


In [5]:
#punto 2
#Preprocesamiento por remocion de la media para X_train y X_test
def remocion_train_test(X_train,X_test):
    media_train = X_train.mean(axis = 0) # calculamos la media del conjunto de entrenamiento
    desviacion_train = X_train.std(axis = 0) # calculamos la desviacion estándar del conjunto de entrenamiento
    
    X_train_rem = (X_train - media_train)/desviacion_train # preprocesamos la matriz de entrenamiento
    X_test_rem = (X_test - media_train)/desviacion_train # preprocesamos la matriz de validacion o test
    
    return X_train_rem, X_test_rem

#Preprocesamiento por escalamiento para X_train y X_test
def escalamiento_train_test(X_train,X_test):
    minimo_train = X_train.min(axis=0)
    maximo_train = X_train.max(axis=0)
    
    X_train_preprocesado = (X_train - minimo_train)/(maximo_train - minimo_train)
    X_test_preprocesado = (X_test - minimo_train)/(maximo_train - minimo_train)
    
    return X_train_preprocesado, X_test_preprocesado

#Preprocesamiento por normalizacion L1 para X_train y X_test
def normalizar_train_test(X_train,X_test): #definimos la función para la norma L1
    norma = np.linalg.norm(X_train,ord=1,axis=1) #calculamos el valor de la norma con la función np.linalg.norm. El orden 
                                            #indica qué tipo de norma es y "axis=1" nos dice que la calculamos sobre las filas
    filas = np.shape(X_train)[0]
    columnas = np.shape(X_train)[1]
    norma1 = np.repeat(norma,columnas) #como tenemos 13 normas en total (1 por cada fila), repetimos cada una tres veces
    norma1 = np.resize(norma,(filas,columnas)) #reorganizamos las normas para que nos queden del mismo tamaño de X_train
    X_train_normalizado = X_train/norma1 #dividimos punto a punto los datos entre la norma
    
    filas = np.shape(X_test)[0]
    columnas = np.shape(X_test)[1]
    norma2 = np.repeat(norma,columnas)
    norma2 = np.resize(norma,(filas,columnas)) #reorganizamos las normas para que nos queden del mismo tamaño de X_test
    X_test_normalizado = X_test/norma2 #dividimos punto a punto los datos entre la norma
    
    return X_train_normalizado, X_test_normalizado #retornamos la matriz normalizada

#X_train y X_test usando preprocesamiento por remocion de la media
X_train_remocion, X_test_remocion = remocion_train_test(X_train, X_test)
print(type(X_train_remocion))

#X_train y X_test usando preprocesamiento por escalamiento
X_train_escalamiento, X_test_escalamiento = escalamiento_train_test(X_train, X_test)

#X_train y X_test usando preprocesamiento por normalización L1
X_train_normalizar, X_test_normalizar = normalizar_train_test(X_train, X_test)

<class 'numpy.ndarray'>


In [6]:
#punto 3
#clasificadores para normalizar, escalamiento y remocion respectivamente
clasificador11 = clasificador6 = clasificador1 = LogisticRegression(C=10.0, random_state = 0) 
clasificador12 = clasificador7 = clasificador2 = LogisticRegression(C=269.0, random_state = 0)
clasificador13 = clasificador8 = clasificador3 = LogisticRegression(C=670.0, random_state = 0)
clasificador14 = clasificador9 = clasificador4 = LogisticRegression(C=700.0, random_state = 0)
clasificador15 = clasificador10 = clasificador5 = LogisticRegression(C=1000.0, random_state = 0)
'''
#clasificadores para escalamiento
LogisticRegression(C=10.0, random_state = 0) 
LogisticRegression(C=269.0, random_state = 0)
LogisticRegression(C=670.0, random_state = 0)
LogisticRegression(C=700.0, random_state = 0)
 LogisticRegression(C=1000.0, random_state = 0)

#clasficadores para normalizar
LogisticRegression(C=10.0, random_state = 0) 
LogisticRegression(C=269.0, random_state = 0)
LogisticRegression(C=670.0, random_state = 0)
LogisticRegression(C=700.0, random_state = 0)
LogisticRegression(C=1000.0, random_state = 0)
'''

'\n#clasificadores para escalamiento\nLogisticRegression(C=10.0, random_state = 0) \nLogisticRegression(C=269.0, random_state = 0)\nLogisticRegression(C=670.0, random_state = 0)\nLogisticRegression(C=700.0, random_state = 0)\n LogisticRegression(C=1000.0, random_state = 0)\n\n#clasficadores para normalizar\nLogisticRegression(C=10.0, random_state = 0) \nLogisticRegression(C=269.0, random_state = 0)\nLogisticRegression(C=670.0, random_state = 0)\nLogisticRegression(C=700.0, random_state = 0)\nLogisticRegression(C=1000.0, random_state = 0)\n'

In [7]:
#punto 4
#entrenamiento para remocion
clasificador1.fit(X_train_remocion,y_train)
clasificador2.fit(X_train_remocion,y_train)
clasificador3.fit(X_train_remocion,y_train)
clasificador4.fit(X_train_remocion,y_train)
clasificador5.fit(X_train_remocion,y_train)

#entrenamiento para escalamiento
clasificador6.fit(X_train_escalamiento,y_train)
clasificador7.fit(X_train_escalamiento,y_train)
clasificador8.fit(X_train_escalamiento,y_train)
clasificador9.fit(X_train_escalamiento,y_train)
clasificador10.fit(X_train_escalamiento,y_train)

#entrenamiento para normalizar
clasificador11.fit(X_train_normalizar,y_train)
clasificador12.fit(X_train_normalizar,y_train)
clasificador13.fit(X_train_normalizar,y_train)
clasificador14.fit(X_train_normalizar,y_train)
clasificador15.fit(X_train_normalizar,y_train)

LogisticRegression(C=1000.0, class_weight=None, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='ovr', n_jobs=1, penalty='l2', random_state=0,
          solver='liblinear', tol=0.0001, verbose=0, warm_start=False)

In [8]:
#Punto 5
#Realizamos el test/validación de los datos y  
#comparamos las etiquetas de la base de datos con las que encontró el clasificador para así fijar el error de clasificación

#Validación para remocion
y_pred1 = clasificador1.predict(X_test_remocion)
y_pred2 = clasificador2.predict(X_test_remocion)
y_pred3 = clasificador3.predict(X_test_remocion)
y_pred4 = clasificador4.predict(X_test_remocion)
y_pred5 = clasificador5.predict(X_test_remocion)

#Validación para escalamiento
y_pred6 = clasificador6.predict(X_test_escalamiento)
y_pred7 = clasificador7.predict(X_test_escalamiento)
y_pred8 = clasificador8.predict(X_test_escalamiento)
y_pred9 = clasificador9.predict(X_test_escalamiento)
y_pred10 = clasificador10.predict(X_test_escalamiento)

#Validación para normalizar
y_pred11 = clasificador11.predict(X_test_normalizar)
y_pred12 = clasificador12.predict(X_test_normalizar)
y_pred13 = clasificador13.predict(X_test_normalizar)
y_pred14 = clasificador14.predict(X_test_normalizar)
y_pred15 = clasificador15.predict(X_test_normalizar)

In [9]:
#punto 6

print('Remocion \n')

print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred1).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred2).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred3).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred4).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred5).sum())

print('\nEscalamiento \n')

print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred6).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred7).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred8).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred9).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred10).sum())

print('\nNormalizacion L1 \n')

print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred11).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred12).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred13).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred14).sum())
print('Las muestras mal clasificadas fueron %d' % (y_test != y_pred15).sum())

Remocion 

Las muestras mal clasificadas fueron 5
Las muestras mal clasificadas fueron 4
Las muestras mal clasificadas fueron 4
Las muestras mal clasificadas fueron 4
Las muestras mal clasificadas fueron 5

Escalamiento 

Las muestras mal clasificadas fueron 18
Las muestras mal clasificadas fueron 15
Las muestras mal clasificadas fueron 16
Las muestras mal clasificadas fueron 16
Las muestras mal clasificadas fueron 16

Normalizacion L1 

Las muestras mal clasificadas fueron 9
Las muestras mal clasificadas fueron 8
Las muestras mal clasificadas fueron 7
Las muestras mal clasificadas fueron 7
Las muestras mal clasificadas fueron 5


In [ ]:
#punto 6
print('La mejor combinación de método de preprocesamiento con combinaciones de C fue Remoción de la media')